### PR TITLE
Update array API tests without adding complex number support

### PR DESCRIPTION
**Pull request by cbourjau**
This PR updates the array API test suite to the latest upstream version ~~plus a minor custom fix that is in the process of being [upstreamed](https://github.com/data-apis/array-api-tests/pull/283)~~ (upstream merged the fix). The fix allows us to run the test suite despite currently not having support for complex numbers, yet (see #32 for more information on adding those). 

The `skips.txt` file contains failures that cause segfaults in the onnxruntime. `xfails.txt` contains test names which fail for other reasons. We should triage both of these files.

**Edit:** Upated the description to reflect the latest circumstance of this PR

### CODE REVIEW

**Comment by adityagoel4512** [comment]
Does this PR supersede https://github.com/Quantco/ndonnx/pull/33?

**Comment by cbourjau** [comment]
> Does this PR supersede #33?

Yes! Sorry for the confusion!

File: skips.txt [hunk 1]
```diff
@@ -28,80 +59,118 @@ array_api_tests/test_has_names.py::test_has_names[linalg-vecdot]
...
+array_api_tests/test_operators_and_elementwise_functions.py::test_acos
+array_api_tests/test_operators_and_elementwise_functions.py::test_acosh
+array_api_tests/test_operators_and_elementwise_functions.py::test_asin
+array_api_tests/test_operators_and_elementwise_functions.py::test_asinh
+array_api_tests/test_operators_and_elementwise_functions.py::test_atan
+array_api_tests/test_operators_and_elementwise_functions.py::test_atanh
+array_api_tests/test_operators_and_elementwise_functions.py::test_clip
+array_api_tests/test_operators_and_elementwise_functions.py::test_conj
+array_api_tests/test_operators_and_elementwise_functions.py::test_copysign
+array_api_tests/test_operators_and_elementwise_functions.py::test_cos
```
Comment (adityagoel4512) [inline skips.txt:85]: These are quite a lot of additions. Is there something systematic that started failing?

Comment (cbourjau) [reply skips.txt:85]: Yes, I also just noticed these. It is a hypothesis error. Possibly locally.

Comment (adityagoel4512) [reply skips.txt:85]: I'm a little confused by what you meant with "possibly locally". Does this mean that some tests fail on your personal machine but not in CI?

Comment (cbourjau) [reply skips.txt:43]: There are now fewer skips than there used to be when you first made the comment, but still too many. The test failures are due to divergences for extreme input values.

**Comment by cbourjau** [comment]
Brief update:

The latest test suite runs quite fast, allowing us to run on many more hypothesis examples than previously (100 instead of 2 at the time of writing). This did surface some further issues. In particular, I discovered that we can trigger segfaults in `onnx` from the `test_roll` ([example](https://github.com/Quantco/ndonnx/actions/runs/10360835846/job/28680032281?pr=37#step:5:459)) and `test_broadcast_arrays` (example will follow when seen on CI) tests.

**Comment by adityagoel4512** [comment]
> Brief update:
> 
> The latest test suite runs quite fast, allowing us to run on many more hypothesis examples than previously (100 instead of 2 at the time of writing). This did surface some further issues. In particular, I discovered that we can trigger segfaults in `onnx` from the `test_roll` ([example](https://github.com/Quantco/ndonnx/actions/runs/10360835846/job/28680032281?pr=37#step:5:459)) and `test_broadcast_arrays` ([example](https://github.com/Quantco/ndonnx/actions/runs/10368662200/job/28702759863?pr=37#step:5:1183)) tests.

Nice, thanks! If it's faster now it would be great to only have `skips.txt` exclusively contain any tests that segfault upstream rather than assertion fails.

**Comment by cbourjau** [comment]
The full list of failures as of this writing can be found in the CI of this [commit](https://github.com/Quantco/ndonnx/pull/37/commits/d25968cd8f5c7ecac41e79e45f455526c98cb10a).
As suggested, the failures that may cause segfaults are in `skips.txt`; all others are in `xfails.txt`. Nonetheless, I didn't configure the test suite to hard-fail on `XPass`es (i.e. xfails that pass unexpectedly) since some of them are flakey even with the higher number of examples.

File: xfails.txt [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+array_api_tests/test_special_cases.py::test_unary[signbit(x_i is +infinity) -> False]
+array_api_tests/test_special_cases.py::test_unary[signbit(x_i is -0) -> True]
+array_api_tests/test_special_cases.py::test_unary[signbit(x_i is -NaN) -> True]
+array_api_tests/test_special_cases.py::test_unary[signbit(x_i is -infinity) -> True]
+array_api_tests/test_special_cases.py::test_unary[sinh(x_i is +0) -> +0]
+array_api_tests/test_special_cases.py::test_unary[sinh(x_i is +infinity) -> +infinity]
+array_api_tests/test_special_cases.py::test_unary[sinh(x_i is -0) -> -0]
+array_api_tests/test_special_cases.py::test_unary[sinh(x_i is -infinity) -> -infinity]
+array_api_tests/test_special_cases.py::test_unary[sinh(x_i is NaN) -> NaN]
+array_api_tests/test_special_cases.py::test_unary[sqrt(x_i < 0) -> NaN]
```
Comment (adityagoel4512) [inline xfails.txt:168]:
```suggestion
```

File: xfails.txt [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+array_api_tests/test_linalg.py::test_tensordot
+array_api_tests/test_linalg.py::test_vecdot
+array_api_tests/test_manipulation_functions.py::test_concat
+array_api_tests/test_manipulation_functions.py::test_expand_dims
+array_api_tests/test_manipulation_functions.py::test_moveaxis
+array_api_tests/test_manipulation_functions.py::test_repeat
+array_api_tests/test_manipulation_functions.py::test_squeeze
+array_api_tests/test_manipulation_functions.py::test_tile
+array_api_tests/test_manipulation_functions.py::test_unstack
+array_api_tests/test_operators_and_elementwise_functions.py::test_acosh
```
Comment (adityagoel4512) [inline xfails.txt:79]:
```suggestion
```

File: xfails.txt [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+array_api_tests/test_operators_and_elementwise_functions.py::test_asinh
```
Comment (adityagoel4512) [inline xfails.txt:80]:
```suggestion
```

File: xfails.txt [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+array_api_tests/test_operators_and_elementwise_functions.py::test_atan2
+array_api_tests/test_operators_and_elementwise_functions.py::test_bitwise_right_shift[__irshift__(x1, x2)]
+array_api_tests/test_operators_and_elementwise_functions.py::test_bitwise_right_shift[__rshift__(x1, x2)]
+array_api_tests/test_operators_and_elementwise_functions.py::test_bitwise_right_shift[bitwise_right_shift(x1, x2)]
```
Comment (adityagoel4512) [inline xfails.txt:84]:
```suggestion
```

**Review by adityagoel4512 — COMMENTED** [review]
There are a few XPASSes.

File: xfails.txt [hunk 6]
```diff
@@ -0,0 +1,173 @@
+array_api_tests/test_constants.py::test_newaxis
+array_api_tests/test_creation_functions.py::test_asarray_arrays
```
Comment (adityagoel4512) [inline xfails.txt:2]:
```suggestion
```

File: xfails.txt [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+array_api_tests/test_has_names.py::test_has_names[manipulation-moveaxis]
+array_api_tests/test_has_names.py::test_has_names[manipulation-repeat]
+array_api_tests/test_has_names.py::test_has_names[manipulation-tile]
+array_api_tests/test_has_names.py::test_has_names[manipulation-unstack]
+array_api_tests/test_inspection_functions.py::test_array_namespace_info
+array_api_tests/test_inspection_functions.py::test_array_namespace_info_dtypes
+array_api_tests/test_linalg.py::test_matrix_transpose
+array_api_tests/test_linalg.py::test_tensordot
+array_api_tests/test_linalg.py::test_vecdot
+array_api_tests/test_manipulation_functions.py::test_concat
```
Comment (adityagoel4512) [inline xfails.txt:72]:
```suggestion
```

File: xfails.txt [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+array_api_tests/test_operators_and_elementwise_functions.py::test_acosh
+array_api_tests/test_operators_and_elementwise_functions.py::test_asinh
+array_api_tests/test_operators_and_elementwise_functions.py::test_atan2
+array_api_tests/test_operators_and_elementwise_functions.py::test_bitwise_right_shift[__irshift__(x1, x2)]
+array_api_tests/test_operators_and_elementwise_functions.py::test_bitwise_right_shift[__rshift__(x1, x2)]
+array_api_tests/test_operators_and_elementwise_functions.py::test_bitwise_right_shift[bitwise_right_shift(x1, x2)]
+array_api_tests/test_operators_and_elementwise_functions.py::test_clip
+array_api_tests/test_operators_and_elementwise_functions.py::test_conj
+array_api_tests/test_operators_and_elementwise_functions.py::test_copysign
+array_api_tests/test_operators_and_elementwise_functions.py::test_cos
```
Comment (adityagoel4512) [inline xfails.txt:88]:
```suggestion
```

File: xfails.txt [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+array_api_tests/test_operators_and_elementwise_functions.py::test_minimum
+array_api_tests/test_operators_and_elementwise_functions.py::test_pow[__ipow__(x, s)]
+array_api_tests/test_operators_and_elementwise_functions.py::test_pow[__ipow__(x1, x2)]
+array_api_tests/test_operators_and_elementwise_functions.py::test_pow[__pow__(x, s)]
+array_api_tests/test_operators_and_elementwise_functions.py::test_pow[__pow__(x1, x2)]
+array_api_tests/test_operators_and_elementwise_functions.py::test_pow[pow(x1, x2)]
+array_api_tests/test_operators_and_elementwise_functions.py::test_real
+array_api_tests/test_operators_and_elementwise_functions.py::test_signbit
+array_api_tests/test_operators_and_elementwise_functions.py::test_sinh
+array_api_tests/test_operators_and_elementwise_functions.py::test_tan
```
Comment (adityagoel4512) [inline xfails.txt:101]:
```suggestion
```

File: xfails.txt [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+array_api_tests/test_special_cases.py::test_unary[sinh(x_i is +0) -> +0]
+array_api_tests/test_special_cases.py::test_unary[sinh(x_i is +infinity) -> +infinity]
+array_api_tests/test_special_cases.py::test_unary[sinh(x_i is -0) -> -0]
+array_api_tests/test_special_cases.py::test_unary[sinh(x_i is -infinity) -> -infinity]
+array_api_tests/test_special_cases.py::test_unary[sinh(x_i is NaN) -> NaN]
+array_api_tests/test_special_cases.py::test_unary[sqrt(x_i < 0) -> NaN]
+array_api_tests/test_statistical_functions.py::test_cumulative_sum
+array_api_tests/test_statistical_functions.py::test_prod
+array_api_tests/test_statistical_functions.py::test_sum
+array_api_tests/test_utility_functions.py::test_all
```
Comment (adityagoel4512) [inline xfails.txt:172]:
```suggestion
```

**Comment by cbourjau** [comment]
> There are a few XPASSes.

I'm afraid some of them are just flaky. I generated these files from observed failures.

File: pixi.toml [hunk 11]
```diff
@@ -43,10 +43,18 @@ hypothesis = "*"
...
 mypy = ">=1.10.0,<1.11"
 ndindex = "*"
+
 [feature.test.tasks]
 test = "pytest"
 test-coverage = "pytest --cov=ndonnx --cov-report=xml --cov-report=term-missing"
-arrayapitests = "ARRAY_API_TESTS_MODULE=ndonnx pytest --max-examples 2 api-coverage-tests/array_api_tests/ --json-report --json-report-file=api-coverage-tests.json -n auto -vv --skips-file=skips.txt"
+
+[feature.test.tasks.arrayapitests]
+cmd = "pytest --max-examples 100 api-coverage-tests/array_api_tests/  -v -rfX --json-report --json-report-file=api-coverage-tests.json -n auto --disable-deadline --disable-extension linalg --skips-file=skips.txt --xfails-file=xfails.txt --hypothesis-seed=0"
```
Comment (cbourjau) [inline pixi.toml:52]: Note the seed! This makes things much easier for CI, but we miss out on the "fuzzing" we would otherwise get over time.

**Comment by cbourjau** [comment]
I am now running the array-api-tests using a fixed seed. Thus, we should no longer get flaky XPASSes. I think this is now ready to merge even if we xfail a lot of tests liberally. We should address these in follow up PRs, IMHO.

**Comment by cbourjau** [comment]
Apparently the tests are still a bit flaky :( The XFAILS you just added were passing previously: https://github.com/Quantco/ndonnx/actions/runs/10677594899/job/29593026155#step:5:1263

**Comment by adityagoel4512** [comment]
> Apparently the tests are still a bit flaky :( The XFAILS you just added were passing previously: https://github.com/Quantco/ndonnx/actions/runs/10677594899/job/29593026155#step:5:1263

I'm not sure how much marginal value there is in 100 iterations of each test compared to around 20. I'd rather make it 20 and have super snappy CI.

**Review by adityagoel4512 — APPROVED** [review]
I reduced the number of repetitions to 16 in the interest of having sub 10 minute CI. I think we can increase it once we've pruned down the list to catch more obscure edge cases but there is a reasonable amount to do as is. The added `xfails` are a little concerning but we can come back to them in a future PR.

File: pixi.toml (hunk 11 above)
Comment (cbourjau) [reply pixi.toml]: Turns out the seed didn't save us from flakiness after all...

**Pull request merged**